File: Problems/1647_Minimum_Deletions_to_Make_Character_Frequencies_Unique.py
class Solution:
    def minDeletions(self, s: str) -> int:
        freq_table = [0]*26
        for c in s:
            freq_table[ord(c)-ord('a')] += 1
        freq_table.sort(reverse=True)
        freq_set = set()
        
        delete_cnt = 0
        for freq in freq_table:
            if freq==0:
                break
            while freq in freq_set:
                freq -= 1
                delete_cnt += 1
            if freq:
                freq_set.add(freq)
        return delete_cnt

if __name__ == '__main__':
    sol = Solution()
    s = "bbcebab"
    print(sol.minDeletions(s))

File: Problems/1647_Minimum_Deletions_to_Make_Character_Frequencies_Unique.cpp
#include<string>
#include<vector>
#include<set>
#include<algorithm>

using namespace std;

class Solution {
public:
    int minDeletions(string s) {
        vector<int> freq_table(26, 0);
        for(auto c: s){
            freq_table[c-'a']++;
        }
        sort(freq_table.begin(), freq_table.end(), greater<int>());
        set<int> freq_set;
        int delete_cnt = 0;
        for(auto freq: freq_table){
            if(!freq) break;
            while(freq_set.count(freq)){
                freq--;
                delete_cnt++;
            }
            if(freq) freq_set.insert(freq);
        }
        return delete_cnt;
    }
};